### PR TITLE
Implementation of the debug_storageRangeAt method

### PR DESCRIPTION
**Pull request by novosandara**
# Description

Implementation of the debug_storageRangeAt method

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [x] I have tested this code manually

### Manual tests

# Documentation update

# Additional comments

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -773,6 +773,38 @@ func (j *jsonRPCHub) GetIteratorDumpTree(block *types.Block, opts *state.DumpInf
...
+
+	blockProposer := types.BytesToAddress(block.Header.Miner)
+
+	transition, err := j.BeginTxn(parentHeader.StateRoot, block.Header, blockProposer)
+	if err != nil {
+		return fmt.Errorf("failed to begin transaction: %w", err)
+	}
+
+	tx := block.Transactions[txIndex]
+	if _, err := transition.Apply(tx); err != nil {
```
Comment (oliverbundalo) [inline server/server.go]: IMO we should execute all transactions with index <= txIndex in order to reach requested state (like we did in TraceTxn for instance).

File: state/txn.go [hunk 2]
```diff
@@ -149,6 +149,49 @@ func (txn *Txn) GetDumpTree(dumpObject *Dump, opts *DumpInfo, deleteEmptyObjects
 	return nextKey, nil
 }
 
+// StorageRangeAt returns the storage at the given block height and transaction index.
+func (txn *Txn) StorageRangeAt(storageRangeResult *StorageRangeResult, addr *types.Address,
+	keyStart []byte, maxResult int) error {
+	object, exists := txn.getStateObject(*addr)
+	if !exists {
```
Comment (oliverbundalo) [inline state/txn.go:156]: I am not sure if we should consider this as an error.

File: server/server.go [hunk 3]
```diff
@@ -797,9 +797,15 @@ func (j *jsonRPCHub) StorageRangeAt(storageRangeResult *state.StorageRangeResult
 		return fmt.Errorf("failed to begin transaction: %w", err)
 	}
 
-	tx := block.Transactions[txIndex]
-	if _, err := transition.Apply(tx); err != nil {
-		return fmt.Errorf("failed to apply transaction %d: %w", txIndex, err)
+	for idx, tx := range block.Transactions {
+		// Execute transactions without tracer until reaching the target transaction
```
Comment (oliverbundalo) [inline server/server.go]: Just delete "without tracer" from the comment

File: jsonrpc/debug_endpoint.go [hunk 4]
```diff
@@ -816,7 +816,7 @@ func (d *Debug) GetAccessibleState(from, to BlockNumber) (interface{}, error) {
 					}
 				}
 
-				return 0, fmt.Errorf("'from' and 'to' block numbers must be different")
+				return 0, fmt.Errorf("no accessible state found on the block %d", start)
```
Comment (oliverbundalo) [inline jsonrpc/debug_endpoint.go]: Consider changing 'on' with 'in':
no accessible state found in the block

File: jsonrpc/debug_endpoint_test.go [hunk 5]
```diff
@@ -1521,7 +1521,7 @@ func TestGetAccessibleState(t *testing.T) {
 				},
 			},
 
-			returnErr: "'from' and 'to' block numbers must be different",
+			returnErr: "no accessible state found on the block",
```
Comment (oliverbundalo) [inline jsonrpc/debug_endpoint_test.go]: Consider changing 'on' with 'in':
no accessible state found in the block

**Review by oliverbundalo — APPROVED** [review]

**Review by Stefan-Ethernal — APPROVED** [review]

**Pull request merged**